scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -391,6 +439,80 @@ object OptimizationRules {
         WithOnComplete(pipe, ComposedOnComplete(fn0, fn1))
     }
   }
+  /**
+   * a.map(f).flatMap(g) == a.flatMap { x => g(f(x)) }
+   * a.flatMap(f).map(g) == a.flatMap { x => f(x).map(g) }
+   *
+   * This is a rule you may want to apply after having
```

Comment:
In Dagon rule optimization does it support this notion? I.e. do you run each rule to stability or do you run all the rules in a loop?

Reply:
Rules don’t, but you can apply a sequence of rules to do this (although, I forgot to use that API in this PR. Using it will clean it up a bit)